link-grammar/parse/preparation.c

```diff
@@ -29,16 +29,22 @@
 
 /**
  * Set c->nearest_word to the nearest word that this connector could
- * possibly connect to.  The connector *might*, in the end,
- * connect to something more distant, but this is the nearest
- * one that could be connected.
+ * possibly connect to.
+ * The connector *might*, in the end, connect to something more distant,
+ * but this is the nearest one that could be connected.
+ * Also recalculate length_limit to be the farthest word number that could
+ * be connected.
  */
-static int set_dist_fields(Connector * c, size_t w, int delta)
+static int set_dist_fields(Connector * c, size_t w, int delta, int w_clamp)
 {
 	int i;
 	if (c == NULL) return (int) w;
-	i = set_dist_fields(c->next, w, delta) + delta;
+	i = set_dist_fields(c->next, w, delta, w_clamp) + delta;
 	c->nearest_word = i;
```

Comment:
This seems to be a simple recursive call. If it's performance critical, then it could be converted into a loop.

(It's almost a tail call; does gcc ever convert tail calls into loops?)

Reply:
gcc may convert tail calls into loops, but it depends on many things and inspecting the result is needed in order to find out.
In our case, `set_dist_fields()` is not converted to loop, and the first X calls are not even getting inlined (as I falsely hoped).

In the various batch profiling it takes 0.4-0.7% of the total CPU, and this percentage will get enlarged (by much) when further optimizations to the rest of the code are implemented. So eventually converting it to a loop (and optimizing it further) may a good idea.